src/js/mailserver.js

```diff
@@ -0,0 +1,150 @@
+//
+// For specific examples, look at the tests in this project.
+
+"use strict";
+var fluid      = fluid || require("infusion");
+var gpii       = fluid.registerNamespace("gpii");
+var namespace  = "gpii.test.mail.smtp";
+var mailServer = fluid.registerNamespace(namespace);
+
+var simplesmtp = require("simplesmtp");
+var fs         = require("fs");
+
+mailServer.defaultMailHandler = function (that, connection) {
+    console.log("You have not registered a custom message handling callback, messages will not be processed further.");
+    console.log("The most recent mail message was saved to '" + that.model.messageFile + "'...");
+};
+
+mailServer.handleStartData = function (that, connection){
+    var timestamp = (new Date()).getTime();
+    that.applier.change('messageFile', that.options.config.outputDir + "/message-" + timestamp + ".txt");
+    connection.saveStream = fs.createWriteStream(that.model.messageFile);
+};
+
+mailServer.handleData = function (that, connection, chunk){
+    connection.saveStream.write(chunk);
+};
+
+mailServer.handleDataReady = function( that, connection, callback){
+    connection.saveStream.end();
+
+    // Execute our payload here once the message is complete and ready for review
+    if (that.model.mailHandler && typeof that.model.mailHandler === "function") {
+        that.model.mailHandler(that, connection);
+    }
+    else {
+        console.log("No mail handler is configured, all mail will be saved and ignored.");
+    }
+
+    callback(null, that.options.config.queueId);
+};
+
+mailServer.init = function (that) {
+    that.simplesmtp = simplesmtp.createServer(that.options.config);
+
+    that.simplesmtp.on("startData", that.handleStartData);
+    that.simplesmtp.on("data",      that.handleData);
+    that.simplesmtp.on("dataReady", that.handleDataReady);
+
+    that.applier.change("mailHandler", that.defaultMailHandler);
+};
+
+// Convenience function to get rid of any custom mail handler and revert to the default.
+mailServer.reset = function(that, callback) {
+    that.applier.change("mailHandler", that.defaultMailHandler);
+    if (callback) { callback(); }
+};
+
+mailServer.start = function (that, callback) {
+    console.log("Starting test mail server on port " + that.options.config.port + "....");
+    that.simplesmtp.listen(that.options.config.port, callback);
+};
+
+mailServer.stop = function(that, callback) {
+    try {
+        that.simplesmtp.end(callback);
+    }
+    catch (e) {
+        console.log("The SMTP server thinks it was already stopped.  I don't care as long as it's no longer running.");
+    }
+};
+
+fluid.defaults(namespace, {
+    gradeNames: ["fluid.standardRelayComponent", "autoInit"],
+    "config": {
+        "SMTPBanner":           "Test Mail Server",
+        "queueID":              "TESTMAIL",
+        "ignoreTLS":            true,
+        "disableDNSValidation": true,
+        "outputDir":            "/tmp",
+        "port":                 4025
+    },
+    "model": {
```

Comment:
As before, I don't see a convincing use-case for allowing these pieces of configuration to change during the lifetime of the component, although in this case you have supplied infrastructure for triggering changes to the latter. In general, the framework supplies sufficient means for overriding configuration "before the fact" to cover the requirements of flexibility - and model material should be more convincingly model-like (capable of being transmitted as JSON). You can simplify this component and move this material to top-level configuration and remove the tests and functionality which allow it to be dynamic. If the user genuinely wants to replace an implementation of a mail server in a running system they can use the existing standard schemes for destroying and recreating a component in-place in the tree.